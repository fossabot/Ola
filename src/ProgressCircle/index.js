import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const progressCircleValue = (value) => {
  value = Math.min(100, Math.max(0, value))
  const diameter = Math.PI * (62 * 2)
  return ((100 - value) / 100) * diameter
}

const showBusyContent = (busy) => {
  if(busy) {
    const busyIsNode = React.isValidElement(busy)
    const busyIsString = typeof(busy) === 'string'
    return busyIsNode || busyIsString
  }
  return false
}

const ProgressCircle = ({ variant, value, description, busy, label, className, ...props }) => {

  const circleStyle = {
    strokeDashoffset: progressCircleValue(value)
  }

  return (
    <div className={cx('ola_progressCircle', variant && `is-${variant}`, {'is-busy': busy}, className)} {...props}>
      { !busy && label && <span className="ola_progressCircle-label">{label}</span> }
      { !busy && (value !== null) &&  <strong className="ola_progressCircle-value">{value}<span>%</span></strong> }
      { !busy && (value !== null) && description && <div className="ola_progressCircle-description">{description}</div> }
      { showBusyContent(busy) && <span className="ola_progressCircle-busy">{busy}</span> }
      <svg className="ola_progressCircle-circle" width="140" height="140" viewBox="0 0 140 140">
        <circle r="62" cx="70" cy="70" className="ola_progressCircle-circle-background"></circle>
        <circle r="62" cx="70" cy="70" className="ola_progressCircle-circle-value" style={circleStyle}></circle>
      </svg>
    </div>
  )
}


ProgressCircle.defaultProps = {
  variant: null,
  value: null,
  description: null,
  label: null,
  busy: false,
  className: null
}

ProgressCircle.propTypes = {
  /** Tag variants */
  variant: PT.oneOf(['error', 'success', 'warning']),
  /** Value of progress */
  value: PT.number,
  /** Description of progress */
  description: PT.oneOfType([
    PT.string, 
    PT.node
  ]),
  /** Busy or loading */
  busy: PT.oneOfType([
    PT.string,
    PT.bool
  ]),
  /** Extra className */
  className: PT.string,
  /** Label */
  label: PT.string
}

export default ProgressCircle
