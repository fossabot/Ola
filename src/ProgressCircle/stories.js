import React from 'react'

import create from '../create'
import { number, radios, boolean } from '@storybook/addon-knobs'

import ProgressCircle from './'

create('ProgressCircle')
  .add('Viewer', () => (
    <figure>
      <ProgressCircle
        variant={radios('Variant', ['null', 'error', 'success', 'warning'])}
        value={number('Value', 20)}
        busy={boolean('Busy', false)} />
    </figure>
  ))
  