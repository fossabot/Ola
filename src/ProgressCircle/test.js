import React from 'react'
import ProgressCircle from './'
import renderer from 'react-test-renderer'

it('Default ProgressCircle', () => {
  const tree = renderer
    .create(<ProgressCircle />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle with error variant', () => {
  const tree = renderer
    .create(<ProgressCircle variant="error" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle with success variant', () => {
  const tree = renderer
    .create(<ProgressCircle variant="success" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle with warning variant', () => {
  const tree = renderer
    .create(<ProgressCircle variant="warning" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle with value', () => {
  const tree = renderer
    .create(<ProgressCircle value={20} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle Busy', () => {
  const tree = renderer
    .create(<ProgressCircle busy={true} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ProgressCircle with extraClass', () => {
  const tree = renderer
    .create(<ProgressCircle extraClass='extraclass' />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
