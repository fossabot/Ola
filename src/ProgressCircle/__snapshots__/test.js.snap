// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default ProgressCircle 1`] = `
<div
  className="ola_progressCircle"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle Busy 1`] = `
<div
  className="ola_progressCircle is-busy"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with error variant 1`] = `
<div
  className="ola_progressCircle is-error"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with extraClass 1`] = `
<div
  className="ola_progressCircle extraclass"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with label and value 1`] = `
<div
  className="ola_progressCircle"
>
  <span
    className="ola_progressCircle-label"
  >
    Your score
  </span>
  <strong
    className="ola_progressCircle-value"
  >
    75
    <span>
      %
    </span>
  </strong>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 97.38937226128358,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with success variant 1`] = `
<div
  className="ola_progressCircle is-success"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with value 1`] = `
<div
  className="ola_progressCircle"
>
  <strong
    className="ola_progressCircle-value"
  >
    20
    <span>
      %
    </span>
  </strong>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 311.6459912361075,
        }
      }
    />
  </svg>
</div>
`;

exports[`ProgressCircle with warning variant 1`] = `
<div
  className="ola_progressCircle is-warning"
>
  <svg
    className="ola_progressCircle-circle"
    height="140"
    viewBox="0 0 140 140"
    width="140"
  >
    <circle
      className="ola_progressCircle-circle-background"
      cx="70"
      cy="70"
      r="62"
    />
    <circle
      className="ola_progressCircle-circle-value"
      cx="70"
      cy="70"
      r="62"
      style={
        Object {
          "strokeDashoffset": 389.55748904513433,
        }
      }
    />
  </svg>
</div>
`;
